File: cart-app/android/app/src/main/kotlin/com/devworld/shop/ui/products/ProductListViewModel.kt
//
//  ProductListViewModel.kt
//  Shop
//
//  Created by Sushant Verma on 15/8/2026 for [/dev/world 2026](https://devworld.au/)
//

package com.devworld.shop.ui.products

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.devworld.shop.bridge.dto.Product
import com.devworld.shop.repo.ShopRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Snapshot of the Products screen's state: the current list plus the available/selected filter and sort. */
data class ProductListUiState(
    val products: List<Product> = emptyList(),
    val availableTags: List<String> = emptyList(),
    val availableSortOptions: List<String> = emptyList(),
    val selectedTag: String? = null,
    val selectedSort: String = "popularity",
)

/** View model for the Products screen: holds the current tag filter and sort choice, and
 * re-derives the displayed product list whenever either changes. */
class ProductListViewModel(private val repository: ShopRepository) : ViewModel() {
    private val _uiState = MutableStateFlow(ProductListUiState())
    val uiState: StateFlow<ProductListUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch(Dispatchers.Default) {
            val tags = repository.getTags()
            val sortOptions = repository.getSortOptions()
            val products = repository.getProducts(emptyList(), _uiState.value.selectedSort)
            _uiState.update { it.copy(availableTags = tags, availableSortOptions = sortOptions, products = products) }
        }
    }

    /** Sets the active tag filter (`null` for no filter) and reloads the product list. */
    fun selectTag(tag: String?) {
        _uiState.update { it.copy(selectedTag = tag) }
        reload()
    }

    /** Sets the active sort option and reloads the product list. */
    fun selectSort(sortOption: String) {
        _uiState.update { it.copy(selectedSort = sortOption) }
        reload()
    }

    /** Re-fetches the product list from the repository using the current tag filter and sort option. */
    private fun reload() {
        viewModelScope.launch(Dispatchers.Default) {
            val state = _uiState.value
            val tags = state.selectedTag?.let { listOf(it) } ?: emptyList()
            val products = repository.getProducts(tags, state.selectedSort)
            _uiState.update { it.copy(products = products) }
        }
    }
}

File: cart-app/ios/Shop/Products/ProductListViewModel.swift
//
//  ProductListViewModel.swift
//  Shop
//
//  Created by Sushant Verma on 15/8/2026 for [/dev/world 2026](https://devworld.au/)
//

import Foundation
import Observation
import ShopSDK

/// View model for the Products tab: holds the current tag filter and sort choice, and
/// re-derives the displayed product list whenever either changes.
@MainActor
@Observable
final class ProductListViewModel {
    private let sdk: ShopSDK

    var products: [Product] = []
    var selectedTag: Tag?
    var selectedSort: SortOption = .popularity
    let availableTags: [Tag]
    let availableSortOptions: [SortOption]

    /// Creates the view model, loading available filter/sort choices and the initial product list.
    init(sdk: ShopSDK) {
        self.sdk = sdk
        self.availableTags = sdk.allTags()
        self.availableSortOptions = sdk.allSortOptions()
        reload()
    }

    /// Sets the active tag filter (`nil` for no filter) and reloads the product list.
    func selectTag(_ tag: Tag?) {
        selectedTag = tag
        reload()
    }

    /// Sets the active sort option and reloads the product list.
    func selectSort(_ sort: SortOption) {
        selectedSort = sort
        reload()
    }

    /// Re-fetches `products` from the SDK using the current tag filter and sort option.
    private func reload() {
        let tags = selectedTag.map { [$0] } ?? []
        products = sdk.products(matching: tags, sortedBy: selectedSort)
    }
}
